cmd/fleet/calendar_cron.go

```diff
@@ -0,0 +1,454 @@
+		team.Config.Integrations.GoogleCalendar.WebhookURL,
+		domain,
+		failingHostsWithoutAssociatedEmail,
+		logger,
+	); err != nil {
+		level.Info(logger).Log("msg", "webhook for hosts without associated email", "err", err)
+	}
+
+	return nil
+}
+
+func processCalendarFailingHosts(
+	ctx context.Context,
+	ds fleet.Datastore,
+	userCalendar fleet.UserCalendar,
+	orgName string,
+	hosts []fleet.HostPolicyMembershipData,
+	logger kitlog.Logger,
+) error {
+	for _, host := range hosts {
+		logger := log.With(logger, "host_id", host.HostID)
+		if err := userCalendar.Configure(host.Email); err != nil {
+			return fmt.Errorf("configure user calendar: %w", err)
+		}
+
+		hostCalendarEvent, calendarEvent, err := ds.GetHostCalendarEvent(ctx, host.HostID)
+
+		deletedExpiredEvent := false
+		if err == nil {
+			if calendarEvent.EndTime.Before(time.Now()) {
+				if err := ds.DeleteCalendarEvent(ctx, calendarEvent.ID); err != nil {
+					level.Info(logger).Log("msg", "deleting existing expired calendar event", "err", err)
+					continue // continue with next host
+				}
+				deletedExpiredEvent = true
+			}
+		}
+
+		switch {
+		case err == nil && !deletedExpiredEvent:
+			if err := processFailingHostExistingCalendarEvent(
+				ctx, ds, userCalendar, orgName, hostCalendarEvent, calendarEvent, host,
+			); err != nil {
+				level.Info(logger).Log("msg", "process failing host existing calendar event", "err", err)
+				continue // continue with next host
+			}
+		case fleet.IsNotFound(err) || deletedExpiredEvent:
+			if err := processFailingHostCreateCalendarEvent(
+				ctx, ds, userCalendar, orgName, host,
+			); err != nil {
+				level.Info(logger).Log("msg", "process failing host create calendar event", "err", err)
+				continue // continue with next host
+			}
+		default:
+			return fmt.Errorf("get calendar event: %w", err)
+		}
+	}
+
+	return nil
+}
+
+func processFailingHostExistingCalendarEvent(
+	ctx context.Context,
+	ds fleet.Datastore,
+	calendar fleet.UserCalendar,
+	orgName string,
+	hostCalendarEvent *fleet.HostCalendarEvent,
+	calendarEvent *fleet.CalendarEvent,
+	host fleet.HostPolicyMembershipData,
+) error {
+	updatedEvent, updated, err := calendar.GetAndUpdateEvent(calendarEvent, func() string {
+		return generateCalendarEventBody(orgName, host.HostDisplayName)
+	})
+	if err != nil {
+		return fmt.Errorf("get event calendar on db: %w", err)
+	}
+	if updated {
+		if err := ds.UpdateCalendarEvent(ctx,
+			calendarEvent.ID,
+			updatedEvent.StartTime,
+			updatedEvent.EndTime,
+			updatedEvent.Data,
+		); err != nil {
+			return fmt.Errorf("updating event calendar on db: %w", err)
+		}
+	}
+	now := time.Now()
+	eventInFuture := now.Before(updatedEvent.StartTime)
+	if eventInFuture {
+		// If the webhook status was sent and event was moved to the future we set the status to pending.
+		// This can happen if the admin wants to retry a remediation.
+		if hostCalendarEvent.WebhookStatus == fleet.CalendarWebhookStatusSent {
```

Comment:
I don't understand this corner case.

Reply:
Disclaimer: We can definitely remove it.

But it was useful during my tests (and may prove useful?):

(Assuming host has policies failing.)

1. I move the calendar event to current time. 
2. Run the cron.
3. Webhook is fired.

(Suppose my Tines script that receives the webhook in (3) has a bug or something.)

1. I move the same calendar event in (1) to now again.
2. Run the cron
3. Webhook is fired.

etc. 

Let's later discuss if we want this here or not.